engine: aarch64 support

TrappM opened the pull request:
## Issue

ovirt engine currently does not support aarch64 hosts

## Changes introduced with this PR

Add basic support for aarch64

## Are you the owner of the code you are sending in, or do you have permission of the owner?

Yes



--

We've been working on aarch64 support in ovirt engine and these are the required changes to get
an aarch64 VM started in our dev setup.

It is still work in progress and there are several open topics,
but we want to share the current status, because we are looking for feedback and support.

We've built and tested with a x86_64 container based setup,
a single aarch64 host on CentOS 9 and local storage.
Installed a VM with CentOS 10 aarch64.
Basic storage/ISO/network access works as expected.
VNC based console access with keyboard + mouse is available.

At the moment, aarch64 is configured to be available in 'compatibility version' 4.8
and the OS type is restricted to 'Other OS'/Linux. Most of the platform specific values in the code
(hugepagesize, maxcpu ...) are copied from other architectures and should be validated.

The feature/model detection for aarch64 seems to be incomplete, we haven't found any 'model_' labeling
for aarch64. This might need additional patches in vdsm or qemu. Due to that, the current model detection in this PR
is very basic - just to get it up and running.


Known issues:

QEMU recommends the 'virt' board for aarch64 VMs and you will also need the following vdsm patch on the aarch64 host
   https://github.com/oVirt/vdsm/commit/4127dfc37b95bd0991141e75f26d40746ed5f857

We had to disable the repo 'resilientstorage' because it seems to be unavailabel for aarch64.


Open topics:

- validate the current aarch64 settings
- add feature/cpu/model detection
- add aarch64 specific tests
- migration
- hotplug mem/cpu

dupondje commented on backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/utils/CompatibilityVersionUpdaterTest.java:
```diff
@@ -49,6 +49,7 @@ private static Map<String, Integer> createMaxNumberOfVmCpusMap() {
         maxVmCpusMap.put("s390x", 384);
         maxVmCpusMap.put("x86", 710);
         maxVmCpusMap.put("ppc", 384);
+        maxVmCpusMap.put("aarch64", 384);
```
Same here, check real value

dupondje commented on backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/validator/VmValidatorTest.java:
```diff
@@ -71,6 +71,7 @@ private static Map<String, Integer> createMaxNumberOfVmCpusMap() {
         maxVmCpusMap.put("s390x", 384);
         maxVmCpusMap.put("x86", MAX_NUM_CPUS);
         maxVmCpusMap.put("ppc", 384);
+        maxVmCpusMap.put("aarch64", 384);
```
Same here, check real value

TrappM replied:
test, See previous comment

dupondje commented on backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/utils/ovf/OvfManagerTest.java:
```diff
@@ -85,6 +85,7 @@ private static Map<String, Integer> createMaxNumberOfVmCpusMap() {
         maxVmCpusMap.put("s390x", 384);
         maxVmCpusMap.put("x86", 16);
         maxVmCpusMap.put("ppc", 384);
+        maxVmCpusMap.put("aarch64", 384);
```
Same here, check real value

TrappM replied:
test, see previous comment

dupondje commented on backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/businessentities/ArchitectureType.java:
```diff
@@ -25,6 +27,7 @@ public enum ArchitectureType implements Identifiable {
     /* Guest architecture */
     ppcle(6, ppc);
 
+    public static final int HOTPLUG_MEMORY_FACTOR_ARM_MB = 256;
```
Is this the real value?

TrappM replied:
# lsmem
Memory block size:       128M

dupondje commented on backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/builder/vminfo/LibvirtVmXmlBuilder.java:
```diff
@@ -1077,11 +1093,21 @@ void writeDevices() {
                 }
+            case aarch64:
+                writeInput();
+                // TODO, is there a better way to define the keyboard?
```
Can't we just use writeInput() (maybe slighyly adapt it?)

TrappM replied:
the current implement is sufficient. If there is no other function available, I would not change it I this PR

dupondje reviewed (changes requested):
First of all thanks for this patch. I think this is a great addition! :)

I added some remarks in the code already. Think it's a very good start, just some things we need to improve.
Did you already had a running Arm VM in oVirt? Or not there yet?

I think it might be useful to have some full libvirt xml of the started VM also.
To see if everything is in there, and if we need to make some improvements.

TrappM commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -1271,7 +1271,9 @@ select fn_db_update_config_value('ServerCPUList',
         || '4:IBM z13s, z13:sie,model_z13-base:z13-base:s390x; '
-        || '5:IBM z14:sie,model_z14-base:z14-base:s390x;',
+        || '5:IBM z14:sie,model_z14-base:z14-base:s390x; '
+        || '1:ARM64 V8.0:asimd:host:aarch64; '
```
That's listed in /proc/cpuinfo and reported via vdsm

processor	: 255
BogoMIPS	: 400.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics cpuid asimdrdm
CPU implementer	: 0x43
CPU architecture: 8
CPU variant	: 0x1
CPU part	: 0x0af
CPU revision	: 1

https://developer.arm.com/documentation/dui0801/l/Overview-of-the-Armv8-Architecture/Advanced-SIMD

TrappM commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -176,23 +176,23 @@ select fn_db_add_config_value('HighUtilizationForEvenlyDistribute','75','general
 select fn_db_add_config_value('MaxMemorySlots','16','general');
 select fn_db_add_config_value('HotPlugMemoryMultiplicationSizeMb','256','general');
-select fn_db_add_config_value_for_versions_up_to('HotUnplugMemorySupported', '{"x86":"true","ppc":"true","s390x":"false"}', '4.8');
+select fn_db_add_config_value_for_versions_up_to('HotUnplugMemorySupported', '{"x86":"true","ppc":"true","s390x":"false","aarch64":"false"}', '4.8');
```
same for HotUnplug

TrappM commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -971,8 +971,8 @@ select fn_db_add_config_value_for_versions_up_to('IsDeferringFileVolumePrealloca
-select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"false","ppc":"false","s390x":"false"}', '4.5');
-select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"true","ppc":"true","s390x":"false"}', '4.8');
+select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"false","ppc":"false","s390x":"false","aarch64":"false"}', '4.5');
+select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"true","ppc":"true","s390x":"false","aarch64":"false"}', '4.8');
```
not in scope of this PR

shubhaOracle commented on backend/manager/modules/utils/src/main/java/org/ovirt/engine/core/utils/OsRepositoryImpl.java:
```diff
@@ -58,6 +58,7 @@ public enum OsRepositoryImpl implements OsRepository {
         defaultOsMap.put(ArchitectureType.x86_64, DEFAULT_X86_OS);
         defaultOsMap.put(ArchitectureType.ppc64, DEFAULT_PPC_OS);
         defaultOsMap.put(ArchitectureType.s390x, DEFAULT_S390_OS);
+        defaultOsMap.put(ArchitectureType.aarch64, DEFAULT_AARCH64_OS);
```
The defaultOsMap  should be new HashMap<>(4) ?

TrappM commented on backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/architecture/HasMaximumNumberOfDisks.java:
```diff
@@ -38,6 +38,11 @@ public void runForS390X() {
 
+    @Override
+    public void runForAARCH64() {
+        hasMaximum = VmCommand.MAX_VIRTIO_CCW_DISKS == countDisks(DiskInterface.VirtIO);
```
switched to MAX_IDE_SLOTS without any issue but it might be a good idea to review that

TrappM commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -81,13 +81,13 @@ select fn_db_add_config_value('BlockMigrationOnSwapUsagePercentage','0','general
-select fn_db_add_config_value('ClusterEmulatedMachines','pc-q35-rhel8.4.0,pc-q35-4.1,pc-i440fx-rhel7.6.0,pc-i440fx-2.12,pseries-rhel8.4.0,s390-ccw-virtio-2.12','4.6');
-select fn_db_add_config_value('ClusterEmulatedMachines','pc-q35-rhel8.6.0,pc-q35-4.1,pc-i440fx-rhel7.6.0,pc-i440fx-2.12,pseries-rhel8.4.0,s390-ccw-virtio-2.12','4.7');
-select fn_db_add_config_value('ClusterEmulatedMachines','pc-q35-rhel9.6.0,pc-i440fx-rhel7.6.0,pseries-rhel8.5.0,s390-ccw-virtio-rhel9.4.0','4.8');
+select fn_db_add_config_value_for_versions_up_to('ClusterEmulatedMachines','pc-i440fx-rhel7.3.0,pc-i440fx-2.6,pseries-rhel7.5.0,s390-ccw-virtio-2.8,virt','4.2');
```
switched to virt-rhel9.6.0

TrappM commented on backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/bll/UpdateVmCommandTest.java:
```diff
@@ -171,6 +171,7 @@ private static Map<String, Integer> createMaxNumberOfVmCpusMap() {
         maxVmCpusMap.put("s390x", 384);
         maxVmCpusMap.put("x86", 512);
         maxVmCpusMap.put("ppc", 384);
+        maxVmCpusMap.put("aarch64", 384);
```
test, See next comment

TrappM commented:
Is there anything blocking this PR or can we get it merged?

According to https://www.ovirt.org/develop/dev-process/install-nightly-snapshot.html
- dnf copr enable -y ovirt/ovirt-master-snapshot centos-stream-9
- dnf install -y ovirt-release-master

it would be great to reduce the effort for the enduser and get some feedback from the community.

dupondje reviewed (changes requested):
Sorry for the delay.
We are currently focussing on getting everything build for CentOS 10, which takes most of our time.

Getting this in would be very nice, but I want to make sure it's (almost) perfect before we merge it.
I added some additional comments when I reviewed the code.
Next to that, I also found some old patch the openkylin project did to support ARM on oVirt:
https://gitee.com/openkylin-backup/ovirt-engine/pulls/1/files

It would also be nice if we could get rid of the TODO's before we merge it.
Cause 'TODO' comments are good, but they will be forgotten and it will stay unimplemented,
So I think it would be nice to have them correct before getting merged.

A last point of attention is the CPU type we should pick for ARM and which flags should be used. I don't know on which CPU you are testing this?

Next to that it could be nice if you joined our Matrix channel. Then we can discuss this easily :)

fraenki commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -81,13 +81,13 @@ select fn_db_add_config_value('BlockMigrationOnSwapUsagePercentage','0','general
-select fn_db_add_config_value('ClusterEmulatedMachines','pc-q35-rhel8.4.0,pc-q35-4.1,pc-i440fx-rhel7.6.0,pc-i440fx-2.12,pseries-rhel8.4.0,s390-ccw-virtio-2.12','4.6');
-select fn_db_add_config_value('ClusterEmulatedMachines','pc-q35-rhel8.6.0,pc-q35-4.1,pc-i440fx-rhel7.6.0,pc-i440fx-2.12,pseries-rhel8.4.0,s390-ccw-virtio-2.12','4.7');
-select fn_db_add_config_value('ClusterEmulatedMachines','pc-q35-rhel9.6.0,pc-i440fx-rhel7.6.0,pseries-rhel8.5.0,s390-ccw-virtio-rhel9.4.0','4.8');
+select fn_db_add_config_value_for_versions_up_to('ClusterEmulatedMachines','pc-i440fx-rhel7.3.0,pc-i440fx-2.6,pseries-rhel7.5.0,s390-ccw-virtio-2.8,virt-rhel9.6.0','4.2');
```
Since ARM support is a new feature, I think it would be perfectly fine to require cluster level 4.8.

TrappM commented on frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/vms/UnitVmModel.java:
```diff
@@ -3466,6 +3466,7 @@ private void validateMemoryAlignment(EntityModel<Integer> model) {
         Cluster cluster = dataCenterWithCluster.getCluster();
         ArchitectureType architectureFamily = cluster.getArchitecture().getFamily();
 
+        // TODO, aarch64
```
I don't know, but it would be great if anybody could sched some light on that.

TrappM commented on packaging/ansible-runner-service-project/project/roles/ovirt-host-deploy-vdsm/tasks/packages.yml:
```diff
@@ -9,6 +9,13 @@
     state: present
   when: ansible_architecture == 'x86_64'
 
+- name: Install centos-release-ovirt45 package on aarch64
```
There was no general 'centos-release-ovirt45' dependency defined but we noticed that it is needed on our aarch64 setup. At the moment I would keep it for aarch64 only.

Commit:
engine: aarch64 support

This patch adds basic support for aarch64 hosts.

The 'Compatibility Version' 4.8 provides
 * CPU architecture aarch64
 * CPU type 'ARM64 V8'

CPU model/feature detection is based on the asmid flag.
Several architecture specific settings have initial values and
should be validated.

The intended scope of this patch is to add initial aarch64 support
and a starting point for further aarch64 related patches.

Co-authored-by: Richard Treu <richard.treu@sap.com>
Signed-off-by: Michael Trapp <michael.trapp@sap.com>

drbugfinder-work commented:
Hi @dupondje 

I hope you’re doing well.

It’s been quite a while since the last activity on this PR. In the meantime, my colleague who originally opened it has stepped away from the project and transferred the repositories to me, so I’ll be maintaining this work going forward.

I just wanted to ask whether you think there is anything still missing before this PR can move forward.

Thanks!

cc @JasperB-TeamBlue

JasperB-TeamBlue commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -176,23 +176,23 @@ select fn_db_add_config_value('HighUtilizationForEvenlyDistribute','75','general
+select fn_db_add_config_value_for_versions_up_to('IsMigrationSupported','{"undefined": "true", "x86": "true", "ppc" : "true", "s390x" : "true", "aarch64" : "true"}','4.8');
 -- snapshot support per architecture
-select fn_db_add_config_value_for_versions_up_to('IsMemorySnapshotSupported','{"undefined": "true", "x86": "true", "ppc" : "true", "s390x" : "true"}','4.8');
+select fn_db_add_config_value_for_versions_up_to('IsMemorySnapshotSupported','{"undefined": "true", "x86": "true", "ppc" : "true", "s390x" : "true", "aarch64" : "true"}','4.8');
```
This config value is not used anymore in the code, it was badly removed in `packaging/dbscripts/upgrade/04_05_0300_remove_vdc_option_IsMemorySnapshotSupported.sql` in which the value was dropped. It does still exist in the DB because this command runs after the upgrade scripts which inserts it again, but there are no stubs in the actual configValues class which call upon this value. it has also been marked deprecated in `backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/action/CreateSnapshotForVmParameters.java`.

JasperB-TeamBlue commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -971,8 +972,8 @@ select fn_db_add_config_value_for_versions_up_to('IsDeferringFileVolumePrealloca
 select fn_db_add_config_value_for_versions_up_to('IsPortIsolationSupported', 'true', '4.8');
-select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"false","ppc":"false","s390x":"false"}', '4.5');
-select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"true","ppc":"true","s390x":"false"}', '4.8');
+select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"false","ppc":"false","s390x":"false"', '4.5');
```
why drop the } ??

JasperB-TeamBlue commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -971,8 +972,8 @@ select fn_db_add_config_value_for_versions_up_to('IsDeferringFileVolumePrealloca
-select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"false","ppc":"false","s390x":"false"}', '4.5');
-select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"true","ppc":"true","s390x":"false"}', '4.8');
+select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"false","ppc":"false","s390x":"false"', '4.5');
+select fn_db_add_config_value_for_versions_up_to('TpmDeviceSupported', '{"x86":"true","ppc":"true","s390x":"false","aarch64":"false"}', '4.8');
```
For each add config value statement that is updated, an equal update_config_value statement should be added to ensure that people who already have a functional database with these values in them, get the updated version.

JasperB-TeamBlue commented on packaging/dbscripts/upgrade/pre_upgrade/0000_config.sql:
```diff
@@ -1183,7 +1184,7 @@ select fn_db_update_config_value_for_versions_from_up_to('PredefinedVMProperties
 select fn_db_update_config_value('HotPlugMemorySupported','{"x86":"true","ppc":"true","s390x":"false"}', '4.2');
 select fn_db_update_config_value('HotUnplugMemorySupported','{"x86":"true","ppc":"true","s390x":"false"}','4.2');
-select fn_db_update_config_value_for_versions_from_up_to('MaxNumOfVmCpus', '{"x86":384,"ppc":384,"s390x":384}', '4.2', '4.4');
+select fn_db_update_config_value_for_versions_from_up_to('MaxNumOfVmCpus', '{"x86":384,"ppc":384,"s390x":384', '4.2', '4.4');
```
why drop the closing bracket?

JasperB-TeamBlue commented on frontend/webadmin/modules/gwt-common/src/main/resources/org/ovirt/engine/ui/common/CommonApplicationConstants_zh_CN.properties:
```diff
@@ -730,6 +730,7 @@ rxNetworkVM=vNIC Rx
 rxRate=Rx \u901F\u7387
 rxTotal=\u603B Rx
 s390xChipset=zseries
+aarch64Chipset=arm64
```
It is not needed to translate this in every language if it is the same for each language. If only the english variant exist it will show the english variant by default. It is best to leave translations to happen on our translations portal.

JasperB-TeamBlue commented on backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/builder/vminfo/LibvirtVmXmlBuilder.java:
Is there any reason to safeguard this block from not executing for aarch64?
